### PR TITLE
[EN TasksMode] Added support for merging holiday and time reference if they appear consecutively under TasksMode

### PR DESCRIPTION
**Pull request by Neelisha-saxena**
Merge holiday with time if they are coming in sequence. 
Now for input texts like "Meet me on Christmas at 5 pm". We will have "Christmas at 5 pm" as extracted entity and its resolved value equals to 25/12/22 T17

### CODE REVIEW

**Review by tellarin — CHANGES_REQUESTED** [review]
@Neelisha-saxena Why spread holiday extraction everywhere? Why not do it in merge extractor/parser? Is there a problem doing it that way?
I worry about the performance impact of the current implementation. Can you ran some performance experiments? If there's no significant impact, I'm kind of ok with it.

**Comment by tellarin** [comment]
@Neelisha-saxena, @fxs130430 may be able to help you with perf test settings.

**Comment by Neelisha-saxena** [comment]
> @Neelisha-saxena Why spread holiday extraction everywhere? Why not do it in merge extractor/parser? Is there a problem doing it that way? I worry about the performance impact of the current implementation. Can you ran some performance experiments? If there's no significant impact, I'm kind of ok with it.

I am shifting changes to tasksmode flag, since it's making code slower right now.

File: .NET/Microsoft.Recognizers.Text.DateTime/Arabic/Parsers/ArabicDateTimePeriodParserConfiguration.cs [hunk 1]
```diff
@@ -45,6 +45,9 @@ public ArabicDateTimePeriodParserConfiguration(ICommonDateTimeParserConfiguratio
             DateTimeParser = config.DateTimeParser;
             TimeZoneParser = config.TimeZoneParser;
 
+            HolidayExtractor = config.HolidayExtractor;
+            HolidaytimeParser = config.HolidaytimeParser;
```
[the remark this answers is not in the document]
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Arabic/Parsers/ArabicDateTimePeriodParserConfiguration.cs:49]: Renamed Paser and push the changes

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishCommonDateTimeParserConfiguration.cs [hunk 2]
```diff
@@ -56,13 +57,15 @@ public EnglishCommonDateTimeParserConfiguration(IDateTimeOptionsConfiguration co
...
             DateParser = new BaseDateParser(new EnglishDateParserConfiguration(this));
+            HolidayTimeParser = new BaseHolidayParser(new EnglishHolidayParserConfiguration(this));
             TimeParser = new TimeParser(new EnglishTimeParserConfiguration(this));
             DateTimeParser = new BaseDateTimeParser(new EnglishDateTimeParserConfiguration(this));
             DatePeriodParser = new BaseDatePeriodParser(new EnglishDatePeriodParserConfiguration(this));
             TimePeriodParser = new BaseTimePeriodParser(new EnglishTimePeriodParserConfiguration(this));
             DateTimePeriodParser = new BaseDateTimePeriodParser(new EnglishDateTimePeriodParserConfiguration(this));
 
             DateTimeAltParser = new BaseDateTimeAltParser(new EnglishDateTimeAltParserConfiguration(this));
+
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishCommonDateTimeParserConfiguration.cs:68]: Nit: Please avoid adding such blank lines.

File: .NET/Microsoft.Recognizers.Text.DateTime/Arabic/Extractors/ArabicDateTimeExtractorConfiguration.cs [hunk 3]
```diff
@@ -85,6 +85,7 @@ public ArabicDateTimeExtractorConfiguration(IDateTimeOptionsConfiguration config
             TimePointExtractor = new BaseTimeExtractor(new ArabicTimeExtractorConfiguration(this));
             DurationExtractor = new BaseDurationExtractor(new ArabicDurationExtractorConfiguration(this));
             UtilityConfiguration = new ArabicDatetimeUtilityConfiguration();
+            HolidayPointExtractor = new BaseHolidayExtractor(new ArabicHolidayExtractorConfiguration(this));
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Arabic/Extractors/ArabicDateTimeExtractorConfiguration.cs:88]: What's the difference between HolidayPointExtractor, SingleHolidayExtractor, and HolidayExtractor?
Shouldn't all be the same name and any individual differences just be handled based on config? That's what it looks like as they're always BaseHolidayExtraction created with a language-specific config.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimeParser.cs [hunk 4]
```diff
@@ -34,6 +34,11 @@ public DateTimeParseResult Parse(ExtractResult er, DateObject refTime)
             if (er.Type.Equals(ParserName, StringComparison.Ordinal))
             {
                 var innerResult = MergeDateAndTime(er.Text, referenceTime);
+                if ((!innerResult.Success) && ((config.Options & DateTimeOptions.TasksMode) != 0))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimeParser.cs:37]: Add a blank line here as the two are unrelated.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 5]
```diff
@@ -397,6 +402,10 @@ protected virtual DateTimeResolutionResult ParseSpecificTimeOfDay(string text, D
                     if (!valid)
                     {
                         ers = this.Config.DateExtractor.Extract(afterStr, referenceTime);
+                        if ((Config.Options & DateTimeOptions.TasksMode) != 0)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:405]: Same, add a blank line. These help in review/merge and follow the workflow (as in the original code).

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 6]
```diff
@@ -454,6 +463,11 @@ protected virtual DateTimeResolutionResult ParseSpecificTimeOfDay(string text, D
                 }
 
                 var pr = this.Config.DateParser.Parse(ers[0], referenceTime);
+                if (pr.Value == null && ((Config.Options & DateTimeOptions.TasksMode) != 0))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:466]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 7]
```diff
@@ -602,6 +634,10 @@ private DateTimeResolutionResult ParseDateWithPeriodPrefix(string text, DateObje
                 if (match.Success)
                 {
                     pr = this.Config.DateParser.Parse(dateResult.Last(), referenceTime);
+                    if ((pr.Value == null) && ((Config.Options & DateTimeOptions.TasksMode) != 0))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:637]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 8]
```diff
@@ -613,6 +649,10 @@ private DateTimeResolutionResult ParseDateWithPeriodPrefix(string text, DateObje
                     if (match.Success)
                     {
                         pr = this.Config.DateParser.Parse(dateResult.First(), referenceTime);
+                        if ((pr.Value == null) && ((Config.Options & DateTimeOptions.TasksMode) != 0))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:652]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 9]
```diff
@@ -686,11 +726,20 @@ private DateTimeResolutionResult MergeDateWithSingleTimePeriod(string text, Date
...
+                    {
+                        dateResult.AddRange(this.Config.HolidayExtractor.Extract(trimmedText.Replace(ers[0].Text, string.Empty), referenceTime));
+                    }
+
                     // Try to add TokenBeforeDate if no result is found because it is not always included in the DateTimePeriod extraction
                     // (e.g. "I'll leave on the 17 from 2 to 4 pm" -> "the 17 from 2 to 4 pm")
                     if (dateResult.Count == 0)
                     {
                         dateResult = this.Config.DateExtractor.Extract(Config.TokenBeforeDate + trimmedText.Substring(0, (int)ers[0].Start), referenceTime);
+                        if ((Config.Options & DateTimeOptions.TasksMode) != 0)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:739]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 10]
```diff
@@ -714,6 +763,10 @@ private DateTimeResolutionResult MergeDateWithSingleTimePeriod(string text, Date
                         DateObject pastTime;
 
                         var pr = this.Config.DateParser.Parse(dateResult[0], referenceTime);
+                        if ((pr.Value == null) && ((Config.Options & DateTimeOptions.TasksMode) != 0))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:766]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 11]
```diff
@@ -808,18 +861,31 @@ private DateTimeResolutionResult ParsePureNumberCases(string text, DateObject re
 
                 // Parse following date
                 var dateExtractResult = this.Config.DateExtractor.Extract(trimmedText.Replace(match.Value, string.Empty), referenceTime);
+                if ((Config.Options & DateTimeOptions.TasksMode) != 0)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:864]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 12]
```diff
@@ -808,18 +861,31 @@ private DateTimeResolutionResult ParsePureNumberCases(string text, DateObject re
...
+                {
+                    dateExtractResult.AddRange(this.Config.HolidayExtractor.Extract(trimmedText.Replace(match.Value, string.Empty), referenceTime));
+                }
 
                 // Try to add TokenBeforeDate if no result is found because it is not always included in the DateTimePeriod extraction
                 // (e.g. "I'll leave on the 17 from 2 to 4 pm" -> "the 17 from 2 to 4 pm")
                 if (dateExtractResult.Count == 0)
                 {
                     dateExtractResult = this.Config.DateExtractor.Extract(Config.TokenBeforeDate + trimmedText.Substring(0, match.Index), referenceTime);
+                    if ((Config.Options & DateTimeOptions.TasksMode) != 0)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:874]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 13]
```diff
@@ -808,18 +861,31 @@ private DateTimeResolutionResult ParsePureNumberCases(string text, DateObject re
...
+                    {
+                        dateExtractResult.AddRange(this.Config.HolidayExtractor.Extract(Config.TokenBeforeDate + trimmedText.Substring(0, match.Index), referenceTime));
+                    }
                 }
 
                 DateObject futureDate, pastDate;
                 if (dateExtractResult.Count > 0)
                 {
                     var pr = this.Config.DateParser.Parse(dateExtractResult[0], referenceTime);
+                    if (((Config.Options & DateTimeOptions.TasksMode) != 0) && (pr.Value == null))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:884]: Same.

**Review by tellarin — CHANGES_REQUESTED** [review]
Minor issues left. This can merge soon.

**Comment by tellarin** [comment]
We should look at the perf issue and enable this for all modes later. Ignore it in this PR as it will be only under TasksMode.

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishCommonDateTimeParserConfiguration.cs (hunk 2 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishCommonDateTimeParserConfiguration.cs]: removed blank line in new commit.

File: .NET/Microsoft.Recognizers.Text.DateTime/Arabic/Extractors/ArabicDateTimeExtractorConfiguration.cs (hunk 3 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Arabic/Extractors/ArabicDateTimeExtractorConfiguration.cs]: renamed  HolidayPointExtractor, SingleHolidayExtractor to HolidayExtractor.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 5 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:406]: added blank line.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 6 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:468]: resolved this in new commit.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimeParser.cs (hunk 4 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimeParser.cs:38]: added a blank line and pushed the change.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 7 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:640]: resolved

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 9 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:744]: resolved

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 11 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:871]: resolved

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 13 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:893]: added blank line.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 8 above)
Comment (Neelisha-saxena) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs:656]: resolved

**Review by tellarin — APPROVED** [review]
Very minor changes pending. Merging as-is, but please make the small changes in one of the followup PRs, ok? Thanks.

**Pull request merged**